### PR TITLE
demonstrate that disable of rule fails some times

### PR DESCRIPTION
**Pull request by maggu2810**
After a rule has been executed, it state is set to IDLE.
This is done regardless if the rule has been deactivated before (at the time it has been executed).

@danchom, @marinmitev Can you have a look at the fix?

Fixes: #1757, #2057

### CODE REVIEW

**Comment by maggu2810** [comment]
@chriskn Can you test this one on your machine(s)?

**Comment by maggu2810** [comment]
Updated PR: I think I have found the real cause of the failing module timer tests.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 1]
```diff
@@ -901,28 +901,33 @@ private void removeMissingModuleTypes(Collection<String> moduleTypes) {
...
+                    logger.debug("The rule '{}' is NOT executed, since it has unsatisfied conditions.", uid);
                 }
             } catch (Throwable t) {
                 logger.error("Fail to execute rule '{}': {}", new Object[] { rule.getUID(), t.getMessage() }, t);
             }
 
-            // change state to IDLE
-            setRuleStatusInfo(rule.getUID(), new RuleStatusInfo(RuleStatus.IDLE));
+            // change state to IDLE only if the rule has not been DISABLED.
+            synchronized (this) {
```
Comment (marinmitev) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Please note that this synchronization affects the execution of all rules which in theory should run in parallel and should not depend on each other. It seems better to synchronize with the rule object itself so other threads executing rules should not be blocked

Comment (marinmitev) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: The rest of the code in the commit looks good to me. Thanks for the fix!

Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: The object `this` is used for synchronization in `setRuleStatusInfo`. That is the reason why it has been choosen.
Shouldn't the `get` and `set` methods for an information use the same sync. object?

**Comment by marinmitev** [comment]
ok, I see that 'this' object is used also to synchronize the access to 'statusMap' which holds all rule statuses
I would suggest to @danchom to consider synchronizing with 'statusMap' instead 'this' in such places.
But since he will be back on Wednesday I will open a new issue about the synchronizations

**Comment by maggu2810** [comment]
> I will open a new issue about the synchronizations

@marinmitev Perhaps you have seen that I fixed another usage of the rule status handling (same method, some lines above).
Have you created the issue for @danchom already (before it gets lost)? I haven't realize it.

**Pull request merged**